plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/CloudSdkServerBehaviour.java

```diff
@@ -0,0 +1,224 @@
+import org.eclipse.debug.core.DebugPlugin;
+import org.eclipse.debug.core.IDebugEventSetListener;
+import org.eclipse.debug.core.ILaunch;
+import org.eclipse.debug.core.ILaunchConfigurationWorkingCopy;
+import org.eclipse.debug.core.ILaunchManager;
+import org.eclipse.debug.core.model.IProcess;
+import org.eclipse.wst.server.core.IServer;
+import org.eclipse.wst.server.core.model.ServerBehaviourDelegate;
+
+import com.google.cloud.tools.eclipse.appengine.localserver.Activator;
+import com.google.common.io.ByteStreams;
+
+/**
+ * A {@link ServerBehaviourDelegate} for Google Cloud SDK.
+ */
+public class CloudSdkServerBehaviour extends ServerBehaviourDelegate {
+  private PingThread pingThread;
+  private IDebugEventSetListener processListener;
+
+  @Override
+  public IStatus canStart(String launchMode) {
+    // Check that the port is not in use before can start
+    CloudSdkServer server = CloudSdkServer.getCloudSdkServer(getServer());
+    int port = server.getApiPort();
+    if (!isPortAvailable(port)) {
+      String message = server.getServer().getName() + " has admin port set to " + port
+          + " which is already is in use.\nTo use port " + port + " stop the processes that are using it";
+      Activator.logError(message);
+      return new Status(IStatus.ERROR, Activator.PLUGIN_ID, message);
+    }
+    return Status.OK_STATUS;
+  }
+
+  @Override
+  public void stop(boolean force) {
+    terminate();
+  }
+
+  /**
+   * Notifies that the server started.
+   */
+  public void setServerStarted() {
+    setServerState(IServer.STATE_STARTED);
+  }
+
+  /**
+   * Finalizes preparations to launch server.
+   *
+   * @param launchMode
+   *          the mode in which a server is running
+   */
+  protected void setupLaunch(String launchMode) {
+    setServerRestartState(false);
+    setServerState(IServer.STATE_STARTING);
+    setMode(launchMode);
+    String adminHost = CloudSdkServer.getCloudSdkServer(getServer()).getApiHost();
+    int adminPort = CloudSdkServer.getCloudSdkServer(getServer()).getApiPort();
+
+    // ping server to check for startup
+    try {
+      String url = "http://" + adminHost + ":" + adminPort;
+      pingThread = new PingThread(getServer(), url, -1, this);
+      pingThread.start();
+    } catch (Throwable e) {
+      Activator.logError("Can't ping for GAE Server startup.", e);
+    }
+  }
+
+  /**
+   * Terminates the Cloud SDK server instance, stops the ping thread and removes
+   * debug listener.
+   */
+  protected void terminate() {
+    int serverState = getServer().getServerState();
+    if ((serverState == IServer.STATE_STOPPED) || (serverState == IServer.STATE_STOPPING)) {
+      return;
+    }
+
+    try {
+      setServerState(IServer.STATE_STOPPING);
+      stopDevAppServer(); // sends a "quit" message to port
+      ILaunch launch = getServer().getLaunch();
+      if (launch != null) {
+        launch.terminate();
+      }
+
+      if (pingThread != null) {
+        pingThread.stop();
```

Comment:
probably OK here; but in general stopping a thread is strongly discouraged for safety reasons